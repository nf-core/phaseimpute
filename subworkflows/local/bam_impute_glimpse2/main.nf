include { GLIMPSE2_PHASE                     } from '../../../modules/nf-core/glimpse2/phase'
include { BCFTOOLS_INDEX as BCFTOOLS_INDEX_1 } from '../../../modules/nf-core/bcftools/index'
include { GLIMPSE2_LIGATE                    } from '../../../modules/nf-core/glimpse2/ligate'
include { BCFTOOLS_INDEX as BCFTOOLS_INDEX_2 } from '../../../modules/nf-core/bcftools/index'

workflow BAM_IMPUTE_GLIMPSE2 {

    take:
    ch_input        // channel (mandatory): [ [id], bam, bai ]
    ch_panel        // channel (mandatory): [ [panel, chr], vcf, tbi ]
    ch_chunks       // channel  (optional): [ [panel, chr], region1, region2 ]
    ch_fasta        // channel (mandatory): [ [genome], fa, fai ]

    main:

    ch_versions = Channel.empty()

    // Impute with Glimpse2 without using binary files
    samples_file = Channel.of([[]]).collect()
    gmap_file    = Channel.of([[]]).collect()

    // Create input channel to impute with Glimpse2

    // Join chunks and panel
    ch_chunks_panel = ch_chunks
        .combine(ch_panel, by:0)
        .map{ metaPC, regionin, regionout, panel, index ->
            [["panel": metaPC.id, "chr": metaPC.chr], regionin, regionout, panel, index]
        }

    // Join input and chunks reference
    ch_phase_input = ch_input
        .combine(samples_file)
        .combine(ch_chunks_panel)
        .combine(gmap_file)
        .map{ metaI, bam, bai, samples, metaPC, regionin, regionout, panel, panel_index, gmap ->
            [metaI + metaPC + ["chunk": regionout],
            bam, bai, samples, regionin, regionout, panel, panel_index, gmap]
        }

    // Impute with Glimpse2
    GLIMPSE2_PHASE(ch_phase_input, ch_fasta)
    ch_versions = ch_versions.mix(GLIMPSE2_PHASE.out.versions)

    // Index phased file
    BCFTOOLS_INDEX_1(GLIMPSE2_PHASE.out.phased_variants)
    ch_versions = ch_versions.mix(BCFTOOLS_INDEX_1.out.versions)

    // Ligate all phased files in one and index it
    ligate_input = GLIMPSE2_PHASE.out.phased_variants
        .join( BCFTOOLS_INDEX_1.out.csi )
        .map{ metaIPCR, vcf, index -> [metaIPCR.subMap("id", "panel", "chr"), vcf, index] }
        .groupTuple()

    GLIMPSE2_LIGATE ( ligate_input )
    ch_versions = ch_versions.mix(GLIMPSE2_LIGATE.out.versions )

    BCFTOOLS_INDEX_2 ( GLIMPSE2_LIGATE.out.merged_variants )
    ch_versions = ch_versions.mix( BCFTOOLS_INDEX_2.out.versions )

    // Join imputed and index files
    ch_imputed_vcf_tbi = GLIMPSE2_LIGATE.out.merged_variants
        .join(BCFTOOLS_INDEX_2.out.tbi)
        .map{ metaIPC, vcf, index -> [metaIPC + [tools: "glimpse2"], vcf, index] }

    emit:
    vcf_tbi             = ch_imputed_vcf_tbi    // channel: [ [id, panel, chr, tool], vcf, tbi ]
    versions            = ch_versions           // channel: [ versions.yml ]
}
