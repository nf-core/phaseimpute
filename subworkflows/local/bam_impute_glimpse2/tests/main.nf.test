nextflow_workflow {

    name "Test Subworkflow BAM_IMPUTE_GLIMPSE2"
    script "../main.nf"

    workflow "BAM_IMPUTE_GLIMPSE2"

    tag "subworkflows"
    tag "subworkflows_local"
    tag "subworkflows/bam_impute_glimpse2"
    tag "bam_impute_glimpse2"

    tag "glimpse2"
    tag "glimpse2/phase"
    tag "glimpse2/ligate"
    tag "bcftools"
    tag "bcftools/index"

    test("Impute with glimpse2") {
        when {
            params {
                max_cpus   = 2
                max_memory = '2.GB'
            }
            workflow {
                """
                input[0] = Channel.fromList([
                    [
                        [id: "NA12878"],
                        file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s.bam", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s.bam.bai", checkIfExist:true),
                    ],
                    [
                        [id: "NA19401"],
                        file(params.pipelines_testdata_base_path + "hum_data/individuals/NA19401/NA19401.s.bam", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/individuals/NA19401/NA19401.s.bam.bai", checkIfExist:true),
                    ],
                ])
                input[1] = Channel.of([
                        [id: "1000GP", chr: "22"],
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.s.norel.vcf.gz", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.s.norel.vcf.gz.csi", checkIfExist:true),
                    ]).collect()
                input[2] = Channel.of(
                    [[chr: "22", id: "1000GP"], "chr22:16570065-16597215", "chr22:16570065-16592216"],
                    [[chr: "22", id: "1000GP"], "chr22:16587172-16609999", "chr22:16592229-16609999"]
                )
                input[3] = Channel.of([
                    [id: "GRCh38"],
                    file(params.pipelines_testdata_base_path + "hum_data/reference_genome/GRCh38.s.fa.gz", checkIfExist:true),
                    file(params.pipelines_testdata_base_path + "hum_data/reference_genome/GRCh38.s.fa.gz.fai", checkIfExist:true),
                ]).collect()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() },
                { assert snapshot(workflow.out.vcf_tbi.collect{ path(it[1]).vcf.summary }).match("Summary") }
            )
        }
    }

    test("Impute with glimpse2 mixed") {
        when {
            params {
                max_cpus   = 2
                max_memory = '2.GB'
            }
            workflow {
                """
                input[0] = Channel.of([
                    [id: "allIndividuals"],
                    [file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s.bam", checkIfExist:true),
                    file(params.pipelines_testdata_base_path + "hum_data/individuals/NA19401/NA19401.s.bam", checkIfExist:true)],
                    [file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s.bam.bai", checkIfExist:true),
                    file(params.pipelines_testdata_base_path + "hum_data/individuals/NA19401/NA19401.s.bam.bai", checkIfExist:true)],
                ])
                input[1] = Channel.of([
                        [id: "1000GP", chr: "22"],
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.s.norel.vcf.gz", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.s.norel.vcf.gz.csi", checkIfExist:true),
                    ]).collect()
                input[2] = Channel.of(
                    [[chr: "22", id: "1000GP"], "chr22:16570065-16597215", "chr22:16570065-16592216"],
                    [[chr: "22", id: "1000GP"], "chr22:16587172-16609999", "chr22:16592229-16609999"]
                )
                input[3] = Channel.of([
                    [id: "GRCh38"],
                    file(params.pipelines_testdata_base_path + "hum_data/reference_genome/GRCh38.s.fa.gz", checkIfExist:true),
                    file(params.pipelines_testdata_base_path + "hum_data/reference_genome/GRCh38.s.fa.gz.fai", checkIfExist:true),
                ]).collect()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() },
                { assert snapshot(workflow.out.vcf_tbi.collect{ path(it[1]).vcf.summary }).match("SummaryMixed") }
            )
        }
    }
}
