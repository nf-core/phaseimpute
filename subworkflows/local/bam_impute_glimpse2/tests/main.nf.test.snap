{
    "Impute with glimpse2": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "NA12878",
                            "panel": "1000GP",
                            "chr": "22",
                            "tools": "glimpse2"
                        },
                        "NA12878.vcf.gz:md5,412f5805ba874bae12d09f818984372e",
                        "NA12878.vcf.gz.tbi:md5,11d96ee2fa0b17d92ad64effced79ef3"
                    ],
                    [
                        {
                            "id": "NA19401",
                            "panel": "1000GP",
                            "chr": "22",
                            "tools": "glimpse2"
                        },
                        "NA19401.vcf.gz:md5,c495afd00ac6c174e8761be47a50a2ac",
                        "NA19401.vcf.gz.tbi:md5,6c38d22c7572b6ec8ecd7576b0b60cb2"
                    ]
                ],
                "1": [
                    "versions.yml:md5,02e20b0099653bc476dac65a92e768c1",
                    "versions.yml:md5,02e20b0099653bc476dac65a92e768c1",
                    "versions.yml:md5,02e20b0099653bc476dac65a92e768c1",
                    "versions.yml:md5,02e20b0099653bc476dac65a92e768c1",
                    "versions.yml:md5,1d8512dd2d5f6649a1111faabd24b399",
                    "versions.yml:md5,1d8512dd2d5f6649a1111faabd24b399",
                    "versions.yml:md5,2bfd212564086adb64cf01741a0d83ad",
                    "versions.yml:md5,2bfd212564086adb64cf01741a0d83ad",
                    "versions.yml:md5,2bfd212564086adb64cf01741a0d83ad",
                    "versions.yml:md5,2bfd212564086adb64cf01741a0d83ad",
                    "versions.yml:md5,7d3bb3301fbcde0b12f553a2ae7d5b56",
                    "versions.yml:md5,7d3bb3301fbcde0b12f553a2ae7d5b56"
                ],
                "vcf_tbi": [
                    [
                        {
                            "id": "NA12878",
                            "panel": "1000GP",
                            "chr": "22",
                            "tools": "glimpse2"
                        },
                        "NA12878.vcf.gz:md5,412f5805ba874bae12d09f818984372e",
                        "NA12878.vcf.gz.tbi:md5,11d96ee2fa0b17d92ad64effced79ef3"
                    ],
                    [
                        {
                            "id": "NA19401",
                            "panel": "1000GP",
                            "chr": "22",
                            "tools": "glimpse2"
                        },
                        "NA19401.vcf.gz:md5,c495afd00ac6c174e8761be47a50a2ac",
                        "NA19401.vcf.gz.tbi:md5,6c38d22c7572b6ec8ecd7576b0b60cb2"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,02e20b0099653bc476dac65a92e768c1",
                    "versions.yml:md5,02e20b0099653bc476dac65a92e768c1",
                    "versions.yml:md5,02e20b0099653bc476dac65a92e768c1",
                    "versions.yml:md5,02e20b0099653bc476dac65a92e768c1",
                    "versions.yml:md5,1d8512dd2d5f6649a1111faabd24b399",
                    "versions.yml:md5,1d8512dd2d5f6649a1111faabd24b399",
                    "versions.yml:md5,2bfd212564086adb64cf01741a0d83ad",
                    "versions.yml:md5,2bfd212564086adb64cf01741a0d83ad",
                    "versions.yml:md5,2bfd212564086adb64cf01741a0d83ad",
                    "versions.yml:md5,2bfd212564086adb64cf01741a0d83ad",
                    "versions.yml:md5,7d3bb3301fbcde0b12f553a2ae7d5b56",
                    "versions.yml:md5,7d3bb3301fbcde0b12f553a2ae7d5b56"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-17T16:14:45.267891348"
    },
    "Impute with glimpse2 mixed": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "allIndividuals",
                            "panel": "1000GP",
                            "chr": "22",
                            "tools": "glimpse2"
                        },
                        "allIndividuals.vcf.gz:md5,292fa79d11120119c2113c1f93868692",
                        "allIndividuals.vcf.gz.tbi:md5,adfb2197e1dcdb86d73c8ce1de316d73"
                    ]
                ],
                "1": [
                    "versions.yml:md5,02e20b0099653bc476dac65a92e768c1",
                    "versions.yml:md5,02e20b0099653bc476dac65a92e768c1",
                    "versions.yml:md5,1d8512dd2d5f6649a1111faabd24b399",
                    "versions.yml:md5,2bfd212564086adb64cf01741a0d83ad",
                    "versions.yml:md5,2bfd212564086adb64cf01741a0d83ad",
                    "versions.yml:md5,7d3bb3301fbcde0b12f553a2ae7d5b56"
                ],
                "vcf_tbi": [
                    [
                        {
                            "id": "allIndividuals",
                            "panel": "1000GP",
                            "chr": "22",
                            "tools": "glimpse2"
                        },
                        "allIndividuals.vcf.gz:md5,292fa79d11120119c2113c1f93868692",
                        "allIndividuals.vcf.gz.tbi:md5,adfb2197e1dcdb86d73c8ce1de316d73"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,02e20b0099653bc476dac65a92e768c1",
                    "versions.yml:md5,02e20b0099653bc476dac65a92e768c1",
                    "versions.yml:md5,1d8512dd2d5f6649a1111faabd24b399",
                    "versions.yml:md5,2bfd212564086adb64cf01741a0d83ad",
                    "versions.yml:md5,2bfd212564086adb64cf01741a0d83ad",
                    "versions.yml:md5,7d3bb3301fbcde0b12f553a2ae7d5b56"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-17T16:15:06.831603159"
    },
    "Summary": {
        "content": [
            [
                "VcfFile [chromosomes=[22], sampleCount=1, variantCount=903, phased=true, phasedAutodetect=true]",
                "VcfFile [chromosomes=[22], sampleCount=1, variantCount=903, phased=true, phasedAutodetect=true]"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-17T15:56:59.174991488"
    },
    "SummaryMixed": {
        "content": [
            [
                "VcfFile [chromosomes=[22], sampleCount=2, variantCount=903, phased=true, phasedAutodetect=true]"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-17T16:15:07.541079198"
    }
}